theseus/optimizer/nonlinear/dcem.py

```diff
@@ -0,0 +1,473 @@
+                if self.normalize:
+                    fX_mu = fX.mean(dim=1).unsqueeze(1)
+                    fX_sigma = fX.std(dim=1).unsqueeze(1)
+                    _fX = (fX - fX_mu) / (fX_sigma + 1e-6)
+                else:
+                    _fX = fX
+
+                if self.n_elite == 1:
+                    # indexes = LML(N=n_elite, verbose=lml_verbose, eps=lml_eps)(-_fX*temp)
+                    indexes = torch.softmax(-_fX * self.temp, dim=1)
+                else:
+                    indexes = LML(N=self.n_elite, verbose=False, eps=self.lml_eps)(
+                        -_fX * self.temp
+                    )
+                indexes = indexes.unsqueeze(2)
+
+            else:
+                indexes_vals = fX.argsort(dim=1)[:, : self.n_elite]
+                # TODO: A scatter would be more efficient here.
+                indexes = torch.zeros(n_batch, self.n_samples, device=device)
+                for j in range(n_batch):
+                    for v in indexes_vals[j]:
+                        indexes[j, v] = 1.0
+                indexes = indexes.unsqueeze(2)
+            # indexes.shape should be (n_batch, n_sample, 1)
+
+            X = X.transpose(0, 1)
+
+            assert indexes.shape[:2] == X.shape[:2]
+            # print("Samples:", X)
+            X_I = indexes * X
+
+            # old_mu = mu.clone().detach()
+
+            mu = torch.sum(X_I, dim=1) / self.n_elite
+            sigma = (
+                (indexes * (X - mu.unsqueeze(1)) ** 2).sum(dim=1) / self.n_elite
+            ).sqrt()
+
+            assert sigma.shape == (n_batch, self.tot_dof)
+
+            with torch.no_grad():
+                err = self.objective.error_squared_norm(self._mu_vec_to_dict(mu)) / 2
+                self._update_info(info, itr, err, converged_indices)
+                if verbose:
+                    print(
+                        f"Nonlinear optimizer. Iteration: {itr+1}. "
+                        f"Error: {err.mean().item()}"
+                    )
+                converged_indices = self._check_convergence(err, info.last_err)
+                info.status[
+                    np.array(converged_indices.cpu().numpy())
+                ] = NonlinearOptimizerStatus.CONVERGED
+
+                # TODO
+                # Doesn't work with lml_eps = 1e-5.
+                # and with lml_eps= 1e-4, gives suboptimal solution
+
+                # if converged_indices.all():
+                #     break  # nothing else will happen at this point
+
+                info.last_err = err
+
+                if end_iter_callback is not None:
+                    end_iter_callback(self, info, mu, itr)
+
+        info.status[
+            info.status == NonlinearOptimizerStatus.START
+        ] = NonlinearOptimizerStatus.MAX_ITERATIONS
+
+        self.objective.update(self._mu_vec_to_dict(mu))
+
+        # return self._mu_vec_to_dict(mu)
+        return itr
+
+    def _solve(self):
```

Comment:
Can maybe rename to `_CEM_step` and switch to passing the previous sigma as an input?

Reply:
yeah